src/storage/mvcc/reader/forward_scanner.rs

```diff
@@ -0,0 +1,426 @@
+                    // Continue iterate next `write`.
+                }
+            }
+
+            self.write_cursor.next(&mut self.statistics.write);
+
+            if !self.write_cursor.valid() {
+                // Key space ended.
+                return Ok(None);
+            }
+            let current_key = self.write_cursor.key(&mut self.statistics.write);
+            if !Key::is_user_key_eq(current_key, user_key.encoded().as_slice()) {
+                // Meet another key.
+                return Ok(None);
+            }
+        }
+    }
+
+    /// Load the value by the given `write`. If value is carried in `write`, it will be returned
+    /// directly. Otherwise there will be a default CF look up.
+    ///
+    /// The implementation is the same as `PointGetter::load_data_by_write`.
+    #[inline]
+    fn load_data_by_write(&mut self, write: Write, user_key: &Key) -> Result<Value> {
+        if self.omit_value {
+            return Ok(vec![]);
+        }
+        match write.short_value {
+            Some(value) => {
+                // Value is carried in `write`.
+                Ok(value)
+            }
+            None => {
+                // Value is in the default CF.
+                self.ensure_default_cursor()?;
+                let value = super::util::near_load_data_by_write(
+                    &mut self.default_cursor.as_mut().unwrap(),
+                    user_key,
+                    write,
+                    &mut self.statistics,
+                )?;
+                Ok(value)
+            }
+        }
+    }
+
+    /// After `self.get()`, our write cursor may be pointing to current user key (if we
+    /// found a desired version), or next user key (if there is no desired version), or
+    /// out of bound.
+    ///
+    /// If it is pointing to current user key, we need to step it until we meet a new
+    /// key. We first try to `next()` a few times. If still not reaching another user
+    /// key, we `seek()`.
+    #[inline]
+    fn move_write_cursor_to_next_user_key(&mut self, current_user_key: &Key) -> Result<()> {
+        for i in 0..SEEK_BOUND {
+            if i > 0 {
+                self.write_cursor.next(&mut self.statistics.write);
+            }
+            if !self.write_cursor.valid() {
+                // Key space ended. We are done here.
+                return Ok(());
+            }
+            {
+                let current_key = self.write_cursor.key(&mut self.statistics.write);
+                if !Key::is_user_key_eq(current_key, current_user_key.encoded().as_slice()) {
+                    // Found another user key. We are done here.
+                    return Ok(());
+                }
+            }
+        }
+
+        // We have not found another user key for now, so we directly `seek()`.
+        // After that, we must pointing to another key, or out of bound.
+        self.write_cursor.seek(
+            &current_user_key.clone().append_ts(0),
```

Comment:
is it valid to have key that ts == 0? If such key is valid, then this line will cause issues. @zhangjinpeng1987

Reply:
`append_ts(0)` is ok here.